uplink/src/collector/systemstats.rs

```diff
@@ -0,0 +1,546 @@
+    }
+
+    fn update(&mut self, proc: &sysinfo::Process, timestamp: u64, sequence: u32) {
+        let sysinfo::DiskUsage { total_written_bytes, written_bytes, total_read_bytes, read_bytes } =
+            proc.disk_usage();
+        self.disk_total_written_bytes = total_written_bytes;
+        self.disk_written_bytes = written_bytes;
+        self.disk_total_read_bytes = total_read_bytes;
+        self.disk_read_bytes = read_bytes;
+        self.cpu_usage = proc.cpu_usage();
+        self.mem_usage = proc.memory();
+        self.timestamp = timestamp;
+        self.sequence = sequence;
+    }
+}
+
+impl Point for Process {
+    fn sequence(&self) -> u32 {
+        self.sequence
+    }
+
+    fn timestamp(&self) -> u64 {
+        self.timestamp
+    }
+}
+
+impl Package for Buffer<Process> {
+    fn topic(&self) -> Arc<String> {
+        self.topic.clone()
+    }
+
+    fn serialize(&self) -> Vec<u8> {
+        serde_json::to_vec(&self.buffer).unwrap()
+    }
+
+    fn anomalies(&self) -> Option<(String, usize)> {
+        self.anomalies()
+    }
+}
+
+struct Processes {
+    sequence: u32,
+    map: HashMap<i32, Process>,
+    stream: Stream<Process>,
+}
+
+impl Processes {
+    fn push(
+        &mut self,
+        id: i32,
+        proc_data: &sysinfo::Process,
+        name: String,
+        timestamp: u64,
+    ) -> Result<(), base::Error> {
+        self.sequence += 1;
+        let proc = self.map.entry(id).or_insert(Process::init(id, name, proc_data.start_time()));
+        proc.update(proc_data, timestamp, self.sequence);
+
+        self.stream.push(proc.clone())
+    }
+}
+
+/// Collects and forward system information such as kernel version, memory and disk space usage,
+/// information regarding running processes, network and processor usage, etc to an IoT platform.
+pub struct StatCollector {
+    /// Handle to sysinfo struct containing system information.
+    sys: System,
+    /// Frequently updated timestamp value, used to ensure idempotence.
+    timestamp: u64,
+    /// System information values to be serialized.
+    system: SystemStats,
+    /// Information about running processes.
+    processes: Processes,
+    /// Individual Processor information.
+    processors: Processors,
+    /// Information regarding individual Network interfaces.
+    networks: Networks,
+    /// Information regarding individual Disks.
+    disks: Disks,
+    /// Uplink configuration.
+    config: Arc<Config>,
+}
+
+impl StatCollector {
+    /// Create and initialize a stat collector
+    pub fn new(config: Arc<Config>, tx: Sender<Box<dyn Package>>) -> Self {
+        let mut sys = System::new();
+        sys.refresh_disks_list();
+        sys.refresh_networks_list();
+
+        let max_buf_size = match config.stats.stream_size {
+            Some(stream_size) => stream_size,
+            None => 10,
+        };
+
+        let mut map = HashMap::new();
+        let stream = Stream::dynamic_with_size(
+            "disk_stats",
```

Comment:
Prefix all the stream names with `uplink_`. We use this prefix to special case stream list in UI